### PR TITLE
Implicitly convert Numpy arrays to Tensors

### PR DESCRIPTION
**Pull request by nilsvu**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@nikwit this seems to work well, so it's a start :)

**Comment by nikwit** [comment]
Cool! Does that mean we can get rid of tensors and datavectors entirely on the python side?

**Comment by nilsvu** [comment]
Bindings still return tensors and DataVectors, and I believe those can't implicitly convert to Numpy arrays. Edit: but we can treat them as such, because they should duck-type as Numpy arrays.

**Comment by nikwit** [comment]
I still want to check if thats not possible somehow, that would be great!

**Comment by nilsvu** [comment]
sure please try

**Comment by nilsvu** [comment]
Maybe specializing `type_caster` has a chance of returning numpy arrays instead of tensors. I think that's how pybind11 implements Eigen types. Not sure if it's worth the effort to try, or if it's actually the right thing to do for tensors.

**Comment by nikwit** [comment]
yes, it looks to be very similar to this: https://stackoverflow.com/questions/42645228/cast-numpy-array-to-from-custom-c-matrix-class-using-pybind11

I'm surprised it is this complicated, after all we only want it to call `np.asarray` whenever we return a tensor.

**Comment by nilsvu** [comment]
Let's do the implicit conversion _from_ numpy arrays first, since it's a big improvement for convenience. Question: enable it for all Tensors, or only scalars and (co-)vectors? For higher-rank tensors (particularly with symmetries) the ordering of indices isn't obvious.

I'm not sure we want to return all tensors as numpy arrays. It's easy enough to convert in Python if you want to: `np.array(tensor)` does it. In many cases you can also duck-type the tensor or DataVector as an array. Tensor objects have useful features like access to component suffixes, symmetries, etc. And we'd have to get reference semantics right so the data isn't copied on the implicit conversion. Overall I see little benefit from implicitly converting tensors to Numpy arrays, but lots of benefit from implicitly converting arrays to tensors (scalars and vectors at least).

**Comment by nilsdeppe** [comment]
I don't think we should go in this direction because the internal ordering of Tensor will not be intuitive for users. Scalar and Vector is fine, we allow this in C++ (array->Tensor, but explicitly), anything higher rank is just asking for trouble. If we end up realizing we should store things in a different order than we currently do, we would break a lot of Python code if they use numpy arrays for e.g. the metric.

**Comment by nikwit** [comment]
I'm happy with the implicit conversion, definitely for scalars and vectors and also for higher ranks. If we decide to change the storage order in C++, then the only thing we should need to change is the buffer protocol so numpy knows how it is stored as far as I understand. On the python side there should not be any confusion about storage indices since we are always using arrays which have a clearly defined memory layout (and no symmetries).

@nilsvu do you want to update #4646 and get rid of the typing in python? If yes, in a follow up PR or directly?

**Comment by nilsdeppe** [comment]
I guess I don't see how it would work for higher rank tensors, but I don't know enough about numpy. For a rank 2 dim 3 symmetric tensor `gamma[1, 2]` might be the underlying location, but then `gamma[2, 1]` would have to alias to `gamma[1, 2]` (or whatever the aliasing we do is). If that can be done in a way that is agnostic to Tensor storage order I guess general bindings can work, I just don't know if it's possible. How would I create such an object in a numpy arary? Can I create symmetric rank 2 numpy arrays? If not, then we simply cannot use numpy arrays for higher ranks because they are fundamentally different.

**Comment by nilsvu** [comment]
I agree and edited the PR so only scalars and vectors can be implicitly converted. I think that's already a big win for convenience.

File: src/DataStructures/Tensor/Python/Tensor.cpp [hunk 1]
```diff
@@ -143,24 +144,32 @@ void bind_tensor_impl(py::module& m, const std::string& name) {  // NOLINT
                  throw std::runtime_error(
                      "Incompatible format: expected a double array.");
                }
-               if (info.ndim != 2) {
+               // 1D arrays are allowed to construct scalars. Higher-rank
+               // tensors require 2D arrays.
+               if (not(info.ndim == 2 or
+                       (info.ndim == 1 and TensorType::rank() == 0))) {
                  throw std::runtime_error(
                      "Tensor data is expected to be 2D with shape (size, "
```
Comment (nikwit) [inline src/DataStructures/Tensor/Python/Tensor.cpp:152]: the error message should tell you explicitly that you only allow scalars and vectors.

File: tests/Unit/DataStructures/Tensor/Python/Test_Tensor.py [hunk 2]
```diff
@@ -35,11 +36,19 @@ def test_construct_from_list(self):
...
         for copy in [True, False]:
             coords = tnsr.I[DataVector, 3, Frame.Inertial](data, copy=copy)
             for i, (a, b) in enumerate(zip(coords, data)):
                 npt.assert_equal(a, b, f"Mismatch at index {i}")
             npt.assert_equal(np.array(coords), data)
+            # Construction of Scalar from 1D array
+            scalar = Scalar[DataVector](data_scalar, copy=copy)
+            npt.assert_equal(np.array(scalar), [data_scalar])
+        # Implicit conversion from Numpy array to Tensor
+        adm_mass_integrand(field=np.random.rand(4),
```
Comment (nikwit) [inline tests/Unit/DataStructures/Tensor/Python/Test_Tensor.py:60]: can you also test Vectors and an error test for tensors?

**Review by nikwit — CHANGES_REQUESTED** [review]
numpy does not support symmetric tensors afaik. My suggestion was to always have full numpy arrays on the python side without symmetries. When converted to C++, we could just check that it satisfies the symmetries of the tensor that the user wants to convert to and copy. But this is fine with me.

File: src/DataStructures/Tensor/Python/Tensor.cpp (hunk 1 above)
Comment (nilsvu) [reply src/DataStructures/Tensor/Python/Tensor.cpp:152]: You can, explicitly, construct higher-rank tensors from numpy arrays. The tensor gives you the index ordering (e.g. with `component_suffix`), so you know how to order the array. Only the implicit conversion is restricted to scalars and vectors. I edited the error message here a bit to include that scalars can be 1D numpy arrays.

File: tests/Unit/DataStructures/Tensor/Python/Test_Tensor.py (hunk 2 above)
Comment (nilsvu) [reply tests/Unit/DataStructures/Tensor/Python/Test_Tensor.py:60]: done

**Review by nilsdeppe — COMMENTED** [review]
I have no additional comments

**Comment by nilsvu** [comment]
@nikwit please check the fixup

**Review by nikwit — DISMISSED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**